{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "966516b1",
   "metadata": {},
   "source": [
    "Instrukcje\n",
    "Zaimplementuj algorytm genetyczny z uwzględnieniem poniższych wymagań:   \n",
    "\n",
    "\tkodowanie binarne bądź dyskretne/ciągłe genotypu   \n",
    "\tparametryzacja długości genotypu, wielkości populacji   \n",
    "\tkrzyżowanie k-punktowe z prawdopodobieństwem pk, losowanie locusów bez zwracania   \n",
    "\tmutacja m-punktowa z prawdopodobieństwem pm, losowanie locusów bez zwracania   \n",
    "\tselekcja:   \n",
    "rankingowa: n najlepszych osobników zastępuje w populacji n najgorszych   \n",
    "koło ruletki z uwzględnieniem funkcji transformującej wartości funkcji przystosowania   \n",
    "\t\twarunek stopu: liczba pokoleń, opcjonalnie docelowa wartość funkcji przystosowania, % najlepszych osobników w populacji   \n",
    "Użyj algorytmu do optymalizacji ANFIS.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a66d4dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np \n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import random\n",
    "from dataclasses import dataclass\n",
    "from typing import List, Tuple\n",
    "\n",
    "\n",
    "@dataclass(frozen=True) # Frozen=True makes instances of this class immutable\n",
    "class Operation:\n",
    "task_id: int #  \n",
    "op_id: int # Every task has 11 operations \n",
    "resource: int\n",
    "duration: int\n",
    "\n",
    "\n",
    "class JobShopData:\n",
    "\"\"\"A dedicated class to hold and provide access to data\"\"\"\n",
    "def __init__(self, path: str):\n",
    "df = pd.read_excel('GA_task.xlsx', header=[0, 1])\n",
    "\n",
    "# A list of all unique operations (genes)\n",
    "self.operations: List[Operation] = []\n",
    "\n",
    "# A structured list of tasks for easy lookup\n",
    "self.tasks: List[List[Operation]] = []\n",
    "\n",
    "# Loop trough task\n",
    "for task_id, task_df in df.groupby(level=0, axis=1): # lvl 0 - ID, level 1 - R/T, axis 0 - rows, axis 1 - columns\n",
    "    # for each task I create a list of its operations\n",
    "    current_task_operation = []\n",
    "\n",
    "    r_col = task_df.xs(\"R\", level=1, axis=1).iloc[:,0] # xs stands for cross section, level 0 is the task ID, level 1 is either 'R' or 'T', axis=1 columns\n",
    "    t_col = task_df.xs(\"T\", level=1, axis=1).iloc[:,0]\n",
    "\n",
    "    # Loop trough the operations for this task\n",
    "    op_id_counter = 0\n",
    "    for resource, time in zip(r_col, t_col):\n",
    "        # Create operation object\n",
    "        op = Operation(\n",
    "            task_id=task_id,\n",
    "            op_id=op_id_counter,\n",
    "            resource=resource,\n",
    "            duration=time\n",
    "        )\n",
    "\n",
    "        # Append new object to the global gene pool\n",
    "        self.operations.append(op)\n",
    "\n",
    "        # Append it to the list for the current task\n",
    "        current_task_operation.append(op)\n",
    "\n",
    "        op_id_counter += 1\n",
    "\n",
    "    # After processing all operations for this task and its list to self.tasks\n",
    "    self.tasks.append(current_task_operation)    \n",
    "\n",
    "self.num_tasks = len(self.tasks)\n",
    "all_resources = {op.resource for op in self.operations}\n",
    "self.num_resources = len(all_resources)\n",
    "self.num_operational_total = len(self.operations)\n",
    "\n",
    "print(f\"Data loaded... \\n\\nNumber of tasks: {self.num_tasks}\\nTootal number of operations: {self.num_operational_total}\\nNumber of resources: {self.num_resources}\\n\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "139460ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataload = JobShopData(path='GA_task.xlsx',)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ac0b4e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Scheduler:\n",
    "    \"\"\"\n",
    "    Calculates the fitness of a single chromosome.\n",
    "    \"\"\"\n",
    "    def __init__(self, data: JobShopData):\n",
    "        \"\"\"\n",
    "        Initializes the Scheduler with the problem data. \n",
    "\n",
    "        Args:\n",
    "            data (JobShopData): The loaded job shop problem data.\n",
    "        \"\"\"\n",
    "        self.data = data\n",
    "    \n",
    "    def calculate_fitness(self, chromosome: List[Operation]) -> float:\n",
    "        \"\"\"\n",
    "        Calculate the makespan for a given chromosome and returns a fitness score.\n",
    "\n",
    "        Args:\n",
    "            chromosome (List[Operation]): A permutation of all operations, representing a schedule's priority list\n",
    "\n",
    "        Returns:\n",
    "            float: The fitness score, calculated as 1.0 makespan.\n",
    "        \"\"\"\n",
    "        # The time when resource will be available\n",
    "        resource_end_times = np.zeros(self.data.num_resources + 1) # Resources are indexed from 1 \n",
    "        \n",
    "        # Times when task are over \n",
    "        task_end_times = np.zeros(self.data.num_tasks + 1)\n",
    "\n",
    "\n",
    "        for op in chromosome:\n",
    "            # Time the task is ready, time of the end of previous operation\n",
    "            task_ready_time = task_end_times[op.task_id] # In first iteration value is zero\n",
    "\n",
    "            # Time when resource is available\n",
    "            resource_ready_time = resource_end_times[op.resource] # In first iteration value is zero\n",
    "\n",
    "            # An operation can only start when task and resource is ready\n",
    "            start_time = max(task_ready_time, resource_ready_time)\n",
    "            end_time = start_time + op.duration\n",
    "\n",
    "            # Update\n",
    "            task_end_times[op.task_id] = end_time\n",
    "            resource_end_times[op.resource] = end_time\n",
    "        \n",
    "        # The longest task took:\n",
    "        total_makespan = np.max(task_end_times) \n",
    "\n",
    "        return 1.0 / total_makespan   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab06ec66",
   "metadata": {},
   "outputs": [],
   "source": [
    "scheduler = Scheduler(dataload)\n",
    "\n",
    "basic_chromosome = dataload.operations\n",
    "\n",
    "fitness = scheduler.calculate_fitness(basic_chromosome)\n",
    "makespan = 1 / fitness\n",
    "\n",
    "print(\"Scheduler test \\n\")\n",
    "print(f\"Fitness of basic is: {fitness} and makespan: {makespan}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c5e8330",
   "metadata": {},
   "outputs": [],
   "source": [
    "class GeneticAlgorithm:\n",
    "    \"\"\"\n",
    "    A genetic algorithm implementation.\n",
    "    \"\"\"\n",
    "    def __init__(self,\n",
    "        fitness_func: callable[[np.ndarray], float],\n",
    "        genotype_length: int,\n",
    "        population_size: int = 100,\n",
    "        gene_type: str = \"continious\",\n",
    "        bounds: tuple[float, float] = (0.0, 1.0),\n",
    "        selection_method: str = \"rank\"):\n",
    "        \"\"\"\n",
    "        Initializes instances of class GeneticAlgorithm\n",
    "        \n",
    "        Args: \n",
    "            fitness_func (callable[[np.ndarray], float]): function accepting a genotype and return float score.\n",
    "            genotype_length (int): Number of genes, parameters\n",
    "            population_size (int): Number of individuals (chromosomes) are in the population\n",
    "            gene_type (str): Available 'continious', 'discrete', 'binary'\n",
    "            bounds (min, max): Constraints for genes\n",
    "            selection_method (str): The method used to select parent from the population\n",
    "        \"\"\"\n",
    "        self.fitness_func = self.fitness_func\n",
    "        self.genotype_length = genotype_length\n",
    "        self.population_size = population_size\n",
    "        self.gene_type = gene_type\n",
    "        self.bounds = bounds\n",
    "\n",
    "        # State\n",
    "        self.population: np.ndarray = None\n",
    "        self.fitness_score: np.ndarray = None\n",
    "        self.best_individual = None\n",
    "        self.best_fitness = -np.inf\n",
    "        \n",
    "        if selection_method == \"roulette\":\n",
    "            self.selection_func = self._selection_roulette\n",
    "        elif selection_method ==\"rank\": \n",
    "            self.selection_func = self._selection_rank\n",
    "        else:\n",
    "            raise ValueError(\"Unknown selection method\")\n",
    "        \n",
    "        self._create_initial_population()\n",
    "\n",
    "    \n",
    "    def _create_initial_population(self):\n",
    "        low, high = self.bounds\n",
    "\n",
    "        # Generate the initial population based on gene type\n",
    "\n",
    "        if self.gene_type == 'continious':\n",
    "            self.population = np.random.uniform(low, high, size=(self.population_size, self.genotype_length))\n",
    "        elif self.gene_type == 'binary':\n",
    "            self.population = np.random.binomial(n=1, p=0.5, size=(self.population_size, self.genotype_length))\n",
    "        elif self.gene_type == 'discrete':\n",
    "            self.population = np.random.randint(low, high + 1, size=(self.population, self.genotype_length))\n",
    "        else:\n",
    "            raise ValueError(f\"Uknown gene type {self.gene_type}\")+\n",
    "        \n",
    "        # Initial evaluation\n",
    "        self.fitness_score = np.array([self.fitness_func(ind) for ind in self.population])\n",
    "        self._update_best()\n",
    "\n",
    "\n",
    "    def _update_best(self):\n",
    "        \"\"\"Updates the global best solution found so far\"\"\"\n",
    "        max_idx = np.argmax(self.fitness_score)\n",
    "        if self.fitness_score[max_idx] > self.best_fitness:\n",
    "            self.best_fitness = self.fitness_score[max_idx]\n",
    "            self.best_individual = self.population[max_idx].copy()\n",
    "\n",
    "    \n",
    "    def _crossover_k_point(self, parent1: np.ndarray, parent2: np.ndarray, k: int, prob_pk: float) -> Tuple[np.ndarray, np.ndarray]:\n",
    "        \"\"\"\n",
    "        Performs Order Crossover\n",
    "\n",
    "        Args:\n",
    "            k (int): Number of cut points\n",
    "            prob_pk (float): Probability of crossover happens\n",
    "        \"\"\"\n",
    "        if random.random() >= prob_pk:\n",
    "            return parent1.copy(), parent2.copy()\n",
    "        \n",
    "        # Draw indices without returning\n",
    "        cut_points = sorted(random.sample(range(1, self.genotype_length), k))\n",
    "\n",
    "        child1 = parent1.copy()\n",
    "        child2 = parent2.copy()\n",
    "\n",
    "        # Swap segments between cut points\n",
    "        swap = False\n",
    "        start = 0\n",
    "\n",
    "        # Add end of array as a virtual cut point\n",
    "        extended_cuts = cut_points + [self.genotype_length]\n",
    "\n",
    "        for cut in extended_cuts:\n",
    "            if swap:\n",
    "                child1[start:cut], child2[start:cut] = parent2[start:cut].copy(), parent1[start:cut].copy()\n",
    "            swap = not swap\n",
    "            start = cut\n",
    "        \n",
    "        return child1, child2\n",
    "\n",
    "    def _mutation_m_points(self, individual: np.ndarray, m: int, prob_pm: float):\n",
    "        \"\"\"\n",
    "        Performs Mutation in m points on a chromosome.\n",
    "\n",
    "        Args:\n",
    "            m (int): Number of genes to mutate\n",
    "            prob_pm (float): Probabilit that mutations happens\n",
    "        \"\"\"\n",
    "\n",
    "        if random.random() >= prob_pm:\n",
    "            return\n",
    "        \n",
    "        # Select m indices to mutate\n",
    "        mutation_indices = random.sample(range(self.genotype_length), m)\n",
    "        low, high = self.bounds\n",
    "\n",
    "        for idx in mutation_indices:\n",
    "            if self.gene_type == 'continious':\n",
    "                individual[idx] = random.uniform(low, high)\n",
    "            elif self.gene_type == 'binary':\n",
    "                individual[idx] = 1 - individual[idx]\n",
    "            elif self.gene_type == 'discrete':\n",
    "                individual[idx] = random.randint(low, high)\n",
    "\n",
    "    \n",
    "    def _selection_roulette(self, transformation_func: callable[[float], float] = None) -> np.ndarray:\n",
    "        \"\"\"\n",
    "        Selects one parent using Roulette Wheel Selection\n",
    "        \n",
    "        Notes:\n",
    "        This function can use optional fitness transformation \n",
    "        \"\"\"\n",
    "        scores = self.fitness_score\n",
    "\n",
    "        # Trasnform \n",
    "        if transformation_func:\n",
    "            adjusted_scores = np.array([transformation_func(s) for s in scores])\n",
    "        else:\n",
    "            adjusted_scores = scores.copy()\n",
    "        \n",
    "        # Handle cases with negative fitness by shifting\n",
    "        if np.min(adjusted_scores) < 0:\n",
    "            adjusted_scores = adjusted_scores - np.min(adjusted_scores) + 1e-6\n",
    "\n",
    "        total = np.sum(adjusted_scores)\n",
    "\n",
    "        if total == 0:\n",
    "            probs = np.ones(len(scores)) / len(scores)\n",
    "        else:\n",
    "            probs = adjusted_scores / total\n",
    "\n",
    "        # Selected parent\n",
    "        chosen_idx = np.random.choice(len(self.population), p=probs)\n",
    "\n",
    "        return self.population[chosen_idx]\n",
    "    \n",
    "\n",
    "    def _selection_rank(self, fitness_scores: list[np.ndarray], n_replace: int):\n",
    "        \"\"\"Selects N best offspring using Rank Selection and replace N worst individuals\"\"\"\n",
    "\n",
    "        ranks = np.argsort(fitness_scores)\n",
    "        total_rank = self.population_size * (self.population_size + 1) / 2 # mathematical shortcut SUM = N * (N + 1) / 2, formula of sum of the first N integers\n",
    "        # total rank is like adding weights 1, 2, 3, 4, 5, 6 \n",
    "        \n",
    "        pick = random.uniform(0, total_rank)\n",
    "        \n",
    "        current_sum = 0\n",
    "        # ranks tells index of the individual at each rank, so rank[0] is the worst, rank[n-1] is best\n",
    "        for i, individual_index in enumerate(ranks):\n",
    "            current_sum += (i + 1)\n",
    "            if current_sum > pick:\n",
    "                return self.population[individual_index]\n",
    "            \n",
    "        return self.population[ranks[-1]]\n",
    "    \n",
    "\n",
    "        def create_child(p1, p2):\n",
    "            child = [None] * size\n",
    "            point1, point2 = random.sample(range(size), 2) # we cant use uniform because it returns float \n",
    "            \n",
    "            # we have to sort this point \n",
    "            start = min(point1, point2)\n",
    "            end = max(point1, point2)\n",
    "\n",
    "            # Copy the slice from parent1 directly into child1\n",
    "            slice_from_parent1 = p1[start:end]\n",
    "            child[start:end] = slice_from_parent1\n",
    "\n",
    "            # Use a set for checking which geners are already in the child\n",
    "            genes_in_slice = set(slice_from_parent1)\n",
    "\n",
    "            # Fill the remaining slots from the second parent (p2)\n",
    "            p2_idx = end # Starting from the end of the slice\n",
    "            child_idx = end\n",
    "            \n",
    "            # Loop unitl the child is full\n",
    "            while None in child:\n",
    "                # If the gene from parent2 is not in the child's slice\n",
    "                if p2[p2_idx] not in genes_in_slice:\n",
    "                    # Add it to the child\n",
    "                    child[child_idx] = p2[p2_idx]\n",
    "\n",
    "                    child_idx = (child_idx + 1) % size # Why this line it's not over loop? because in case when the gen is in the slice, position to fill\n",
    "                    # shouldn't move\n",
    "                p2_idx = (p2_idx + 1) % size\n",
    "\n",
    "            return child\n",
    "\n",
    "        child1 = create_child(parent1, parent2)\n",
    "        child2 = create_child(parent2, parent1)\n",
    "\n",
    "        return child1, child2\n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "   \n",
    "    \n",
    "\n",
    "    def run(self, num_generations: int):\n",
    "        # Create new popultion\n",
    "        self._create_initial_population()\n",
    "\n",
    "        for gen in range(num_generations):\n",
    "            # Evolution\n",
    "            fitness_scores = [self.scheduler.calculate_fitness(chromo) for chromo in self.population]\n",
    "            \n",
    "            # Track best solution\n",
    "            index_of_best_in_gen = np.argmax(fitness_scores) # argmax returns index of max value\n",
    "            if fitness_scores[index_of_best_in_gen] > self.best_fitness:\n",
    "                self.best_fitness = fitness_scores[index_of_best_in_gen]\n",
    "                self.best_chromosome = self.population[index_of_best_in_gen].copy()\n",
    "\n",
    "            # Create nexxt generation\n",
    "            next_generation = []\n",
    "\n",
    "            # Elitism\n",
    "            best_in_gen_chromosome = self.population[index_of_best_in_gen]\n",
    "            next_generation.append(best_in_gen_chromosome.copy())\n",
    "\n",
    "\n",
    "            while len(next_generation) < self.population_size:\n",
    "                # Selection\n",
    "                parent1 = self.selection_func(fitness_scores)\n",
    "                parent2 = self.selection_func(fitness_scores)\n",
    "\n",
    "                # Crossover\n",
    "                if random.random() < self.crossover_rate:\n",
    "                    new_parent1, new_parent2 = self._crossover(parent1, parent2)\n",
    "                else:\n",
    "                    new_parent1, new_parent2 = parent1.copy(), parent2.copy()\n",
    "\n",
    "                # Mutation\n",
    "                self._mutation(new_parent1)\n",
    "                self._mutation(new_parent2)\n",
    "\n",
    "                offspring1 = new_parent1.copy()\n",
    "                offspring2 = new_parent2.copy()\n",
    "\n",
    "                next_generation.append(offspring1)\n",
    "\n",
    "                # Making sure not to cross tthe limit of population size\n",
    "                if len(next_generation) < self.population_size:\n",
    "                    next_generation.append(offspring2)\n",
    "\n",
    "            self.population = next_generation\n",
    "\n",
    "            if (gen + 1) % 10 == 0:\n",
    "                print(f\"Generation {gen + 1}/{num_generations}. Best fitness: {self.best_fitness}\")\n",
    "\n",
    "        return self.best_chromosome, self.best_fitness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6b547af",
   "metadata": {},
   "outputs": [],
   "source": [
    "gen_alg = GeneticAlgorithm(dataload, scheduler)\n",
    "best_chromosome, best_fitness = gen_alg.run(num_generations=5000)\n",
    "makespan = 1 / best_fitness\n",
    "print(f\"Best fitness score is equal:{best_fitness} and makespan: {makespan}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e18ecf66",
   "metadata": {},
   "outputs": [],
   "source": [
    "best_chromosome"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pymc_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
